sql/src/main/java/io/crate/protocols/postgres/ConnectionContext.java

```diff
@@ -343,9 +359,55 @@ public void channelDisconnected(ChannelHandlerContext ctx, ChannelStateEvent e)
     }
 
     private void handleStartupBody(ChannelBuffer buffer, Channel channel) {
-        session = readStartupMessage(buffer);
-        Messages.sendAuthenticationOK(channel);
+        sessionContext = readStartupMessage(buffer);
+        session = sqlOperations.createSession(sessionContext);
+        authenticate(channel);
+    }
+
+    private void authenticate(Channel channel) {
+        InetAddress address = getRemoteAddress(channel);
+        AuthenticationMethod authMethod = authService.resolveAuthenticationType(sessionContext.userName(), address);
+        if (authMethod == null) {
+            String errorMessage = String.format(
+                Locale.ENGLISH,
+                "No valid auth.host_based entry found for host \"%s\", user \"%s\", schema \"%s\"",
+                address.getHostAddress(), sessionContext.userName(), sessionContext.defaultSchema()
+            );
+            Messages.sendAuthenticationError(channel, errorMessage);
+        } else {
+            authMethod.pgAuthenticate(channel, sessionContext)
+                .whenComplete((success, throwable) -> {
+                    if (throwable == null) {
+                        if (success) {
+                            if (LOGGER.isTraceEnabled()) {
+                                LOGGER.trace("Authentication succeeded user \"{}\" and method \"{}\".",
+                                    sessionContext.userName(), authMethod.name());
+                            }
+                            sendReadyForQuery(channel);
+                        } else {
+                            LOGGER.warn("Authentication failed for user \"{}\" and method \"{}\".",
```

Comment:
the error response to the client is sent from inside `pgAuthenticate()`, the consumer here is only a callback when the response was written.
https://github.com/crate/crate/pull/5512/files#diff-449200b54aa09ce9af0635aaab4cc366R41